workspaces/src/types.rs

```diff
@@ -0,0 +1,243 @@
+            ));
+        }
+        // Adapted from https://github.com/near/near-sdk-rs/blob/fd7d4f82d0dfd15f824a1cf110e552e940ea9073/near-sdk/src/environment/env.rs#L819
+
+        // NOTE: We don't want to use Regex here, because it requires extra time to compile it.
+        // The valid account ID regex is /^(([a-z\d]+[-_])*[a-z\d]+\.)*([a-z\d]+[-_])*[a-z\d]+$/
+        // Instead the implementation is based on the previous character checks.
+
+        // We can safely assume that last char was a separator.
+        let mut last_char_is_separator = true;
+
+        for c in account_id.bytes() {
+            let current_char_is_separator = match c {
+                b'a'..=b'z' | b'0'..=b'9' => false,
+                b'-' | b'_' | b'.' => true,
+                _ => {
+                    return Err(ParseAccountError(
+                        ParseErrorKind::Invalid,
+                        account_id.to_string(),
+                    ))
+                }
+            };
+            if current_char_is_separator && last_char_is_separator {
+                return Err(ParseAccountError(
+                    ParseErrorKind::Invalid,
+                    account_id.to_string(),
+                ));
+            }
+            last_char_is_separator = current_char_is_separator;
+        }
+
+        (!last_char_is_separator)
+            .then(|| ())
+            .ok_or_else(|| ParseAccountError(ParseErrorKind::Invalid, account_id.to_string()))
+    }
+}
+
+impl TryFrom<String> for AccountId {
+    type Error = ParseAccountError;
+
+    fn try_from(account_id: String) -> Result<Self, Self::Error> {
+        Self::validate(&account_id)?;
+        Ok(Self(account_id.into()))
+    }
+}
+
+impl From<AccountId> for String {
+    fn from(account_id: AccountId) -> Self {
+        account_id.0.into_string()
+    }
+}
+
+impl From<AccountId> for near_primitives::types::AccountId {
+    fn from(id: AccountId) -> Self {
+        id.0.into_string()
+            .try_into()
+            .expect("Should not fail since this is already validated before")
+    }
+}
+
+impl FromStr for AccountId {
+    type Err = ParseAccountError;
+
+    fn from_str(account_id: &str) -> Result<Self, Self::Err> {
+        Self::validate(account_id)?;
+        Ok(Self(account_id.into()))
+    }
+}
+
+impl fmt::Display for AccountId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// An error occurred when parsing an invalid Account ID with [`AccountId::validate`](crate::AccountId::validate).
+#[derive(Eq, Hash, Clone, Debug, PartialEq)]
+pub struct ParseAccountError(pub(crate) ParseErrorKind, pub(crate) String);
+
+/// A list of errors that occur when parsing an invalid Account ID.
+#[derive(Eq, Hash, Clone, Debug, PartialEq)]
+pub enum ParseErrorKind {
+    TooLong,
+    TooShort,
+    Invalid,
+}
+
+/// Signer that keeps secret key in memory.
+#[derive(Clone)]
+pub struct InMemorySigner {
```

Comment:
Why do we need this type to be public in the first place? None of the examples use it. Can we just supply worker with secret keys? Something like 

```rust
impl<T> Worker<T> {
    pub fn register_account(&mut self, account_id: AccountId, key_pair: (PublicKey, SecretKey))
}
```

?

Reply:
It's because `create_tla` and `create_tla_and_deploy` functions require it. But I think we can just switch those to take in key_pair instead then

Reply:
Actually, there's a couple functions that would require it too, like `delete_account` and `transfer_near` that I think would be more intuitive if it had signer over a key pair.

Reply:
TBH, I feel that these functions' signatures are less than ideal.

```rust
    pub(crate) async fn transfer_near(
        &self,
        signer: &InMemorySigner,
        receiver_id: AccountId,
        amount_yocto: Balance,
    ) -> anyhow::Result<FinalExecutionOutcomeView> {
```

From the user's POV, "signer" doesn't make sense. As a user, I want to transfer from one account id to another account id, so I'd expect `fn transfer(&self, from: AccountId, to: AccountId, amount: Ballance)`. The `Signer`, and especially `InMemory` bit, feel like impl details that leak into the interface. Can we hide this from the user completely?  Would something like this work?

```rust
fn test_transfer(w: Worker) {
   let alice: Account = w.create_account("alice");
   let bob: Account = w.create_account("bob");

   alice.transfer_near(bob.id(), 9001);
}
```

To put in other way, why to we need to expose `Signer` from `dev_generate` in the first place? 

```
async fn dev_generate(&self) -> (AccountId, InMemorySigner) {
```

Can we return an `Account` instead, or stash the signer in some kind of a hash map?

Reply:
> why to we need to expose Signer from dev_generate in the first place?

Originally, just a helper to let us sign transactions more easily, but I changed it to return `SecretKey` instead now since that was good enough to create a signer from.

> Can we return an Account instead, or stash the signer in some kind of a hash map?

Yeah, I was going to add some convenience functions to `Account`s in a future PR. Already have one for `call` here: https://github.com/near/workspaces-rs/pull/33. We could also do what near-api-js does and have a global keystore associated to an account-id, but that might be a bit hard to track when keys are being registered. That was at least my experience when I was looking through the code. @volovyk-s what do you think about adding a `KeyStore`-like object to register accounts and tracks keys?

Reply:
@ChaoticTempest KeyStore is a good pattern, I think we can use it here. Ideally, it should be a part of the unexisting `near-api-rs` library :/
You can check the pattern in `near-api-js` for now.

Reply:
Ok sounds good then. Let's just go with the convenience functions in a later PR then